### PR TITLE
Server side eval support in JS/React SDKs

### PR DESCRIPTION
**Pull request by bryce-fitzsimons**
### Features and Changes

- Closes **(add link to issue here)**

### CODE REVIEW

File: packages/sdk-js/src/GrowthBook.ts [hunk 1]
```diff
@@ -885,7 +893,32 @@ export class GrowthBook<
     else console.log(msg, ctx);
   }
 
+  public enqueueTrackedExperiments(
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:896]: Trying to generalize process of decoupled experiment tracking

File: packages/sdk-js/src/feature-repository.ts [hunk 2]
```diff
@@ -289,8 +301,13 @@ function startAutoRefresh(instance: GrowthBook): void {
       src: null,
       cb: (event: MessageEvent<string>) => {
         try {
-          const json: FeatureApiResponse = JSON.parse(event.data);
-          onNewFeatureData(key, json);
+          if (serverSideEval) {
+            const json: { update: boolean } = JSON.parse(event.data);
+            if (json.update) fetchFeatures(instance);
```
Comment (bttf) [inline packages/sdk-js/src/feature-repository.ts:306]: What happens here? Do we need to call `onNewFeatureData` or similar for changes to take affect?

Comment (bryce-fitzsimons) [reply packages/sdk-js/src/feature-repository.ts:306]: no, `onNewFeatureData` lets us take in a full sdk payload over the SSE message and ingest it. In our case (serverSideEval===true), no payload is coming through, Instead `{ update: true }` comes through, which informs that SDK that a new payload is available. The SDK will then manually fetch that payload (server-side-evaluated).

File: packages/sdk-js/src/GrowthBook.ts (hunk 1 above)
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts:905]: maybe new method to pass in + process trackEvents

File: packages/sdk-js/src/GrowthBook.ts [hunk 3]
```diff
@@ -131,6 +133,10 @@ export class GrowthBook<
     ];
   }
 
+  public getServerSideEval(): boolean {
+    return this._ctx.serverSideEval || false;
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:137]: rename to "remote/proxy eval" or something

File: packages/sdk-js/src/GrowthBook.ts [hunk 4]
```diff
@@ -894,7 +941,29 @@ export class GrowthBook<
...
+  }
+
+  public fireTrackingCalls(data: TrackExperimentData | TrackExperimentData[]) {
+    data = Array.isArray(data) ? data : [data];
+    data.forEach(({ experiment, result }) => {
+      this._track(experiment, result);
+    });
+  }
+
+  public getTrackingCallsQueue(): TrackExperimentData[] {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:958]: we could also call this `getDeferredTrackingCalls()` if we like...

File: packages/sdk-js/src/GrowthBook.ts [hunk 5]
```diff
@@ -106,7 +108,7 @@ export class GrowthBook<
       this._updateAllAutoExperiments();
     }
 
-    if (context.clientKey) {
+    if (context.clientKey && !context.remoteEval) {
       this._refresh({}, true, false);
```
[the remark this answers is not in the document]
Comment (bryce-fitzsimons) [reply packages/sdk-js/src/GrowthBook.ts]: update: looks like is no longer happening. reverted.

File: packages/sdk-js/src/GrowthBook.ts [hunk 6]
```diff
@@ -894,7 +937,29 @@ export class GrowthBook<
     else console.log(msg, ctx);
   }
 
+  public enqueueTrackingCalls(
+    data: TrackExperimentData | TrackExperimentData[]
+  ) {
+    data = Array.isArray(data) ? data : [data];
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:943]: should we bother with this, or just be opinionated about accepting an array only?

File: packages/sdk-js/src/feature-repository.ts [hunk 7]
```diff
@@ -289,8 +297,14 @@ function startAutoRefresh(instance: GrowthBook): void {
       src: null,
       cb: (event: MessageEvent<string>) => {
         try {
-          const json: FeatureApiResponse = JSON.parse(event.data);
-          onNewFeatureData(key, json);
+          if (instance.getRemoteEval()) {
+            // got an "update" SSE message, trigger a refetch
+            const json: { update: boolean } = JSON.parse(event.data);
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/feature-repository.ts:302]: could optionally send the timestamp of lastUpdated, but likely overkill

File: packages/sdk-js/src/GrowthBook.ts [hunk 8]
```diff
@@ -251,6 +305,51 @@ export class GrowthBook<
     return new Map(this._assigned);
   }
 
+  public exportState({
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:308]: save this for a separate PR maybe?

File: packages/sdk-js/src/GrowthBook.ts [hunk 9]
```diff
@@ -899,7 +998,29 @@ export class GrowthBook<
...
+      this._track(experiment, result);
+    });
+  }
+
+  public getTrackingCallsQueue(): TrackExperimentData[] {
+    return this._trackingCallsQueue;
+  }
+
   private _track<T>(experiment: Experiment<T>, result: Result<T>) {
+    if (this._ctx.deferTracking) {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:1020]: keep this as a separate PR relating to hydration

File: packages/sdk-js/src/GrowthBook.ts [hunk 10]
```diff
@@ -899,7 +998,29 @@ export class GrowthBook<
     else console.log(msg, ctx);
   }
 
+  public enqueueTrackingCalls(
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:1001]: not needed?

File: packages/sdk-js/src/GrowthBook.ts [hunk 11]
```diff
@@ -899,7 +998,29 @@ export class GrowthBook<
...
+    data: TrackExperimentData | TrackExperimentData[]
+  ) {
+    data = Array.isArray(data) ? data : [data];
+    this._trackingCallsQueue.push(...data);
+  }
+
+  public fireTrackingCalls(data: TrackExperimentData | TrackExperimentData[]) {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:1008]: needed?

File: packages/sdk-js/src/GrowthBook.ts [hunk 12]
```diff
@@ -899,7 +998,29 @@ export class GrowthBook<
...
+    data = Array.isArray(data) ? data : [data];
+    data.forEach(({ experiment, result }) => {
+      this._track(experiment, result);
+    });
+  }
+
+  public getTrackingCallsQueue(): TrackExperimentData[] {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/GrowthBook.ts:1015]: not needed

File: packages/sdk-js/src/types/growthbook.ts [hunk 13]
```diff
@@ -246,7 +262,10 @@ export type CacheSettings = {
 
 export type ApiHost = string;
 export type ClientKey = string;
-export type RepositoryKey = `${ApiHost}||${ClientKey}`;
+export type UserKey = string;
+export type RepositoryKey =
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/types/growthbook.ts:266]: use string instead (comments good)

File: packages/sdk-js/src/feature-repository.ts [hunk 14]
```diff
@@ -251,12 +250,35 @@ async function refreshInstance(
...
+  const options: RequestInit = remoteEval
+    ? {
+        method: "POST",
+        headers: { "Content-Type": "application/json", ...apiRequestHeaders },
+        body: JSON.stringify({
+          attributes: instance.getAttributes(),
+        }),
+      }
+    : {
+        headers: apiRequestHeaders,
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/feature-repository.ts:276]: double check no weird side effects

File: packages/sdk-js/src/feature-repository.ts [hunk 15]
```diff
@@ -299,30 +327,48 @@ function startAutoRefresh(instance: GrowthBook): void {
       src: null,
       cb: (event: MessageEvent<string>) => {
         try {
-          const json: FeatureApiResponse = JSON.parse(event.data);
-          onNewFeatureData(key, json);
+          if (event.type === "features-updated") {
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/feature-repository.ts:330]: double check that event types are actually being sent (specifically "features")

File: packages/sdk-js/src/types/growthbook.ts [hunk 16]
```diff
@@ -246,7 +262,10 @@ export type CacheSettings = {
...
+  | `${ApiHost}||${ClientKey}` // local eval
+  | `${ApiHost}||${ClientKey}||${UserKey}`; // remote eval
```
Comment (bryce-fitzsimons) [inline packages/sdk-js/src/types/growthbook.ts:269]: todo: test cases
